File: lifecycle/src/lifecycle/manager.cpp
//
// ROS1 Lifecycle - A library implementing the ROS2 lifecycle for ROS1
//
// Copyright 2016,2017 Robert Bosch GmbH
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "lifecycle/manager.h"
#include "lifecycle/lifecycle_model.h"
#include "lifecycle/broadcaster.h"
#include <boost/format.hpp>
#include <ros/callback_queue_interface.h>

namespace ros { namespace lifecycle {

using std::make_pair;
LifecycleManager::LifecycleManager(const ros::NodeHandle& nh) :
        nh_(nh),
        as_(nh, LIFECYCLE_ACTION_NAME, false),
        current_(UNCONFIGURED),
        lm_broadcaster_(nh){
    primary_steps_[make_pair(UNCONFIGURED, CONFIGURE)]      = Configuring;
    primary_steps_[make_pair(UNCONFIGURED, SHUTDOWN)]       = ShuttingDown;
    
    primary_steps_[make_pair(INACTIVE, CLEANUP)]            = CleaningUp;
    primary_steps_[make_pair(INACTIVE, ACTIVATE)]           = Activating;
    primary_steps_[make_pair(INACTIVE, SHUTDOWN)]           = ShuttingDown;
    primary_steps_[make_pair(INACTIVE, ERROR)]              = ErrorProcessing;
        
    primary_steps_[make_pair(ACTIVE, SHUTDOWN)]             = ShuttingDown;
    primary_steps_[make_pair(ACTIVE, DEACTIVATE)]           = Deactivating;
    primary_steps_[make_pair(ACTIVE, ERROR)]                = ErrorProcessing;
        
    secondary_steps_[make_pair(Configuring, SUCCESS)]       = INACTIVE;
    secondary_steps_[make_pair(Configuring, FAILURE)]       = UNCONFIGURED;
    
    secondary_steps_[make_pair(CleaningUp, SUCCESS)]        = UNCONFIGURED; // must not fail
    
    secondary_steps_[make_pair(Activating, SUCCESS)]        = ACTIVE;
    secondary_steps_[make_pair(Activating, FAILURE)]        = INACTIVE;
    
    secondary_steps_[make_pair(ShuttingDown, SUCCESS)]      = FINALIZED; // must not fail

    secondary_steps_[make_pair(Deactivating, SUCCESS)]      = INACTIVE; // must not fail

    secondary_steps_[make_pair(ErrorProcessing, SUCCESS)]   = UNCONFIGURED;
    secondary_steps_[make_pair(ErrorProcessing, FAILURE)]   = FINALIZED;
    
    setTransitionCallback(ERROR, boost::bind(&LifecycleManager::activeEx_cb, this) );

    state_pub_ = nh_.advertise<lifecycle_msgs::Lifecycle>(LIFECYCLE_STATE_TOPIC, true);
    as_.registerGoalCallback(boost::bind(&LifecycleManager::goalCb, this));
}

LifecycleManager::~LifecycleManager() {
    
}

void LifecycleManager::publishTransition(const Transition& transition, const ResultCode& result_code) {
    lifecycle_msgs::Lifecycle msg;
    msg.transition = transition;
    msg.end_state = current_;
    msg.result_code = result_code;
    
    state_pub_.publish(msg);
    lm_broadcaster_.send_lm_event(msg);
}

namespace {
using namespace ros;

	class ActivationCallback : public CallbackInterface {
	public:
		ActivationCallback(LifecycleManager* lcm) : lcm(lcm) {};

		CallbackInterface::CallResult call() {
			try {
				if(lcm->configure()) {
					lcm->activate();
					return Success;
				} else {
					ROS_WARN_STREAM("Configuring " << ros::this_node::getName()
					<< " failed, won't activate");
				}
			} catch(const std::exception& ex) {
				ROS_ERROR_STREAM("Error when enabling "
						<< ros::this_node::getName() << ": " << ex.what());
			} catch(...) {
				ROS_ERROR_STREAM("Unknown error when enabling "
										<< ros::this_node::getName());
			}
			return Invalid;
		}
	private:
		LifecycleManager* lcm;
	};
}

void LifecycleManager::start() {
    as_.start();
    bool lifecycle_enabled = false;
    nh_.param(PARAM_LIFECYCLE_MANAGEMENT, lifecycle_enabled, false);
    if(!lifecycle_enabled) {
    	// schedule a callback that will activate this node in the absence of
    	// a manager
    	nh_.getCallbackQueue()->addCallback(boost::make_shared<ActivationCallback>(this));
    }
}

void LifecycleManager::setTransitionCallback(Transition tr, transitionCb cb) {
    switch(tr) {
        case CONFIGURE:
            callbacks_[make_pair(UNCONFIGURED, tr)] = cb;
            break;
        case ACTIVATE:
            callbacks_[make_pair(INACTIVE, tr)] = cb;
            break;
        case DEACTIVATE:
            callbacks_[make_pair(ACTIVE, tr)] = cb;
            break;
        case SHUTDOWN:
            callbacks_[make_pair(ACTIVE, tr)] = cb;
            callbacks_[make_pair(INACTIVE, tr)] = cb;
            callbacks_[make_pair(UNCONFIGURED, tr)] = cb;
            break;
        case CLEANUP:
            callbacks_[make_pair(INACTIVE, tr)] = cb;
            break;
        case ERROR:
            callbacks_[make_pair(ACTIVE, tr)] = cb;
            callbacks_[make_pair(INACTIVE, tr)] = cb;
            break;
        default:
            assert(false);
    }
}

void LifecycleManager::goalCb() {
    lifecycle_msgs::LifecycleGoalConstPtr goal = as_.acceptNewGoal();
    lifecycle_msgs::LifecycleResult result;
    try {
        if (handleTransition(static_cast<Transition>(goal->transition))) {
            result.end_state = getCurrentState();
            as_.setSucceeded(result, "goal state reached okay");
        } else {
            result.end_state = getCurrentState();
            as_.setAborted(result, "transition failed");
        }
    } catch(const IllegalTransitionException& ite) {
        result.end_state = getCurrentState();
        as_.setAborted(result, "requested transition is not a valid lifecycle transition");
    }
}

bool LifecycleManager::handleTransition(const Transition& transition) {
    PrimaryInput p_input = make_pair(current_, transition);
    bool result;
    result = handlePrimaryStep(p_input);
    try {
        if(result) {
            handleSecondaryStep(make_pair(current_, SUCCESS));
        } else {
            handleSecondaryStep(make_pair(current_, FAILURE));
        }
    } catch (const std::exception& ex) {
        if(handleErrorProcessing(ex)) {
            handleSecondaryStep(make_pair(current_, SUCCESS));
        } else {
            handleSecondaryStep(make_pair(current_, FAILURE));
        }
    }
    
    if(result) {
        publishTransition(transition, SUCCESS);
    } else {
        publishTransition(transition, FAILURE);
    }
    
    return result;
}

bool LifecycleManager::handlePrimaryStep(const PrimaryInput& input) {
    PrimaryStepMap::const_iterator primary_steps_iter = primary_steps_.find(input);
    CallbackMap::iterator cb_iter = callbacks_.find(input);
    if(primary_steps_iter != primary_steps_.end()) {
        current_ = primary_steps_iter->second;
        if(cb_iter != callbacks_.end()) {
            try {
                return cb_iter->second();
            } catch(const std::exception& ex) {
                return handleErrorProcessing(ex);
            }
        } else {
            return true;
        }
    } else {
        throw IllegalTransitionException(input);
    }
}
  
bool LifecycleManager::handleSecondaryStep(const SecondaryInput& input) {
    SecondaryStepMap::const_iterator secondary_steps_iter = secondary_steps_.find(input);
    if(secondary_steps_iter != secondary_steps_.end()) {
        current_ = secondary_steps_iter->second;
        return true;
    } else {
        throw IllegalTransitionException(input);
    }
}

bool LifecycleManager::handleErrorProcessing(const std::exception& ex) {
    current_ = ErrorProcessing;
    try {
        return onError_(ex);
    } catch(const std::exception& ex) {
        return false;
    }
}

void LifecycleManager::setErrorCb(errorCb cb) {
    onError_ = cb;
}

bool LifecycleManager::raiseError(const std::exception& ex) {
    if(LifecycleModel::isPrimaryState(current_)) {
        activeEx_ = ex;
        return handleTransition(ERROR);
    } else {
        return handleErrorProcessing(ex);
    }
}

bool LifecycleManager::configure() {
    return handleTransition(CONFIGURE);
}

bool LifecycleManager::activate() {
    return handleTransition(ACTIVATE);
}

bool LifecycleManager::deactivate() {
    return handleTransition(DEACTIVATE);
}

bool LifecycleManager::shutdown() {
    return handleTransition(SHUTDOWN);
}

bool LifecycleManager::cleanup() {
    return handleTransition(CLEANUP);
}

IllegalTransitionException::IllegalTransitionException(const PrimaryInput& input) {

    message = str(boost::format("IllegalTransitionException state=%1%, transition=%2%") %
                      input.first % input.second);
};
IllegalTransitionException::IllegalTransitionException(const SecondaryInput& input) {

    message = str(boost::format("IllegalTransitionException state=%1%, transition=%2%") %
                      input.first % input.second);
};
IllegalTransitionException::~IllegalTransitionException() throw() { }
const char* IllegalTransitionException::what() const throw() {
    return message.c_str();
}

}}

File: lifecycle_python/src/lifecycle/manager.py
#!/usr/bin/env python

from inspect import isfunction, ismethod, getcallargs

import rospy
import actionlib

from lifecycle_msgs.msg import LifecycleGoal, LifecycleAction, LifecycleResult, Lifecycle
from lifecycle.lifecycle_model import State, Transition, Result_Code, LifecycleModel
from lifecycle.broadcaster import LmEventBroadcaster

LIFECYCLE_ACTION_NAME = "lifecycle"
LIFECYCLE_STATE_TOPIC = "lifecycle_state"
    
class IllegalTransitionException(Exception):
    def __init__(self, arg):
        self.msg = 'IllegalTransitionException current state =' + str(arg[0]) + ', transition requested =' +str(arg[1])
        
class LifecycleCbException(Exception):
    """
    Thrown on invalid callback function
    """
    def __init__(self, msg):
        Exception.__init__(self, msg)

def check_args(fn, *args):
    '''Checks if the argument passed is a function or a method; and also checks for number of arguments'''
    if not isfunction(fn) and not ismethod(fn):
        return False
    try:
        getcallargs(fn, *args)
        return True
    except TypeError as e:
        return False

class LifecycleManager(object):
    def __init__(self, component_fqn):
        self._callbacks = {}
        self._current = State.UNCONFIGURED
        '''Variable to store the exception occurred during ACTIVE state'''
        self._active_ex = Exception('None')
        '''A lambda expression which raises an exception for the ERROR transition call back'''
        self.set_transition_callback(Transition.ERROR, lambda : (_ for _ in ()).throw(self._active_ex))
        self._as = actionlib.SimpleActionServer(component_fqn + "/" + LIFECYCLE_ACTION_NAME, LifecycleAction, self._goal_cb, False)
        self.state_pub_ = rospy.Publisher(component_fqn + "/" + LIFECYCLE_STATE_TOPIC, Lifecycle, queue_size = 10, latch=True)
        self.lm_broadcaster = LmEventBroadcaster(component_fqn);
        
    def __del__(self):
        self._as.__del__()

    def _publish_transition(self, transition, result_code):
        """
        Publishes the Lifecycle events. Whenever the state of the node changes,
        this method is called to publish the transition that triggered the change,
        current state and the result code of the transition
        :param transition: transition requested
        :type transition: Transition
        :param result_code: the result of the transition
        :type result_code: Result_Code
        :return: 
        """
        msg = Lifecycle()
        msg.transition = transition
        msg.end_state = self._current
        msg.result_code = result_code
        
        self.state_pub_.publish(msg)
        self.lm_broadcaster.sendLmEvent(msg)

    def start(self):
        """
        Starts the action server
        :param :
        :return:
        """
        self._as.start()

    def set_transition_callback(self, tr, cb):
        """
        Sets the callback functions of the respective transition
        :param tr : transition 
        :type tr: Transition
        :param cb : callback function
        :type cb : function
        :return:
        """
        if cb is None or not check_args(cb):
            raise LifecycleCbException("The callback %s must be callable with no arguments" %cb)
        for item in LifecycleModel.STATE_TRANSITION_VALIDPAIRS[tr]:
            self._callbacks[item] = cb

    def _goal_cb(self, goal):
        """
        A goal callback function for the action server updates the result of the action
        :param :
        :return:
        """
        result = LifecycleResult()
        try:
            if(self._handle_transition(goal.transition) != 0):
                result.end_state = self.get_current_state()
                self._as.set_succeeded(result, 'goal state reached okay')
            else:
                result.end_state = self.get_current_state()
                self._as.set_aborted(result, 'transition failed')
        except IllegalTransitionException:
            result.end_state = self.get_current_state()
            self._as.set_aborted(result, 'requested transition is not valid lifecycle transition')

    def _handle_transition(self, transition):
        """
        Handles the transition of node from one primary state to another primary state
        :param transition: requested Transition
        :type transition: Transition
        :return result_var: bool value to indicate success or failure
        """
        input_var = (self._current, transition)
        result_var = self._handle_primary_step(input_var)
        try :
            if(result_var != False):
                self._handle_secondary_step((self._current, Result_Code.SUCCESS))
            else:
                self._handle_secondary_step((self._current, Result_Code.FAILURE))
        except Exception as ex:
            if(self._handle_error_processing(ex)):
                self._handle_secondary_step((self._current, Result_Code.SUCCESS))
            else:
                self._handle_secondary_step((self._current, Result_Code.FAILURE))
                
        if(result_var != False):
            self._publish_transition(transition, Result_Code.SUCCESS)
        else :
            self._publish_transition(transition, Result_Code.FAILURE)
            
        return result_var

    def _handle_primary_step(self, input_var):
        """
        handles the first step of the transition i.e. from a primary state to a transitional state
        :param input_var: has the current state and the transition requested
        :type input_var: (State, Transition)
        :return result_var: bool value to indicate success or failure
        """
        try:
            #this statement may throw an exception if key is not found
            self._current = LifecycleModel.PRIMARY_STEPS[input_var]
            try:
                #this statement may throw an exception if key is not found
                cb_func = self._callbacks[input_var]
                try:
                    return cb_func()
                except Exception as ex:
                    return self._handle_error_processing(ex)
            except KeyError:
                return True
        except KeyError:
            raise IllegalTransitionException(input_var)
            
    def _handle_secondary_step(self, input_var):
        """
        Handles the second step of the transition i.e. from a transitional state to a primary state.
        :param input_var: has the current state and the result code
        :type input_var: (State, Result_Code)
        :return result_var: bool value to indicate success or failure
        """
        try:
            self._current = LifecycleModel.SECONDARY_STEPS[input_var]
            return True
        except KeyError:
            raise IllegalTransitionException(input_var)

    def _handle_error_processing(self,ex):
        """
        Handles the ErrorProcessing step of a transition
        :param ex: exception
        :type ex: Exception
        :return : bool value to indicate success or failure
        """
        self._current = State.ErrorProcessing
        try :
            return self.onError_(ex)
        except Exception as ex:
            return False

    def set_error_cb(self, cb):
        if cb is None or not check_args(cb, Exception):
            raise LifecycleCbException("The error callback %s must be callable with one argument" %cb)
        self.onError_ = cb
    
    def _isPrimaryState(self, state):
        """
        checks if the given state is Primary
        :param state: state of the node
        :param type: STATE
        :return : bool value to indicate if the given state is primary state
        """
        return((state == State.UNCONFIGURED) or (state == State.INACTIVE) or (state == State.ACTIVE) or (state == State.FINALIZED))
        
    def raise_error(self, ex):
        """
        Handles the error raised during any state of the node
        :param ex: exception
        :type ex: Exception
        :return : bool value to indicate success or failure
        """
        if (self._isPrimaryState(self._current)):
            self._active_ex = ex;
            return self._handle_transition(Transition.ERROR)
        else:
            return self._handle_error_processing(ex)
    
    def configure(self):
        return self._handle_transition(Transition.CONFIGURE)

    def activate(self):
        return self._handle_transition(Transition.ACTIVATE)

    def deactivate(self):
        return self._handle_transition(Transition.DEACTIVATE)

    def shutdown(self):
        return self._handle_transition(Transition.SHUTDOWN)

    def cleanup(self):
        return self._handle_transition(Transition.CLEANUP)

    def get_current_state(self):
        return self._current
